File: tests/test_pair_lambda.cpp
// The PairLambda variants: cyclic closing and fixed endpoints (#163).
//
// Three things are pinned here that nothing else pins: the VALUE of each
// variant against a hand computation (short lists included, since a
// variable-length List makes n = 0, 1, 2 routine rather than exotic), the
// agreement between incremental and from-scratch evaluation after every list
// move generator, and the `.cbls` round-trip -- where an open chain must still
// serialise to the bytes it always did.

#include "test_helpers.h"

#include <algorithm>
#include <array>
#include <catch2/catch_test_macros.hpp>
#include <catch2/matchers/catch_matchers_floating_point.hpp>
#include <cbls/cbls.h>
#include <cmath>
#include <limits>
#include <numeric>
#include <sstream>
#include <string>
#include <vector>

using namespace cbls;

namespace {

// Asymmetric on purpose: a symmetric cost cannot distinguish a cyclic sum from
// twice an open one on n = 2, nor a tour from its reverse.
double d(int a, int b) {
    return (10.0 * a) + b;
}

double head_cost(int e) {
    return 100.0 + e;
}

double tail_cost(int e) {
    return 1000.0 + e;
}

// Build a one-node model over a List of `capacity` positions and evaluate it at
// `elements`. `elements` may be shorter than the list's capacity: that is how
// the n = 0, 1, 2 cases are reached without a separate model each.
double value_at(int capacity, const std::vector<int32_t>& elements, PairMode mode, bool endpoints) {
    Model m;
    auto lv = m.list_var(capacity, "seq");
    int32_t node = endpoints ? m.pair_lambda_sum(lv, d, head_cost, tail_cost, mode)
                             : m.pair_lambda_sum(lv, d, mode);
    m.minimize(node);
    m.close();
    m.var_mut(vid(lv)).elements = elements;
    full_evaluate(m);
    return m.node_value(node);
}

}  // namespace

TEST_CASE("pair_lambda_sum values for n = 0, 1, 2 and a general n", "[dag][pair_lambda]") {
    SECTION("n = 0 is zero for every variant, endpoints included") {
        REQUIRE(value_at(4, {}, PairMode::Open, false) == 0.0);
        REQUIRE(value_at(4, {}, PairMode::Cyclic, false) == 0.0);
        REQUIRE(value_at(4, {}, PairMode::Open, true) == 0.0);
        REQUIRE(value_at(4, {}, PairMode::Cyclic, true) == 0.0);
    }

    SECTION("n = 1 has no pair, and Cyclic adds none") {
        REQUIRE(value_at(4, {3}, PairMode::Open, false) == 0.0);
        REQUIRE(value_at(4, {3}, PairMode::Cyclic, false) == 0.0);
        // head(3) + tail(3) = 103 + 1003
        REQUIRE(value_at(4, {3}, PairMode::Open, true) == 1106.0);
        REQUIRE(value_at(4, {3}, PairMode::Cyclic, true) == 1106.0);
    }

    SECTION("n = 2 traverses its one edge twice when cyclic") {
        REQUIRE(value_at(4, {1, 2}, PairMode::Open, false) == 12.0);           // d(1,2)
        REQUIRE(value_at(4, {1, 2}, PairMode::Cyclic, false) == 12.0 + 21.0);  // + d(2,1)
        // d(1,2) + head(1) + tail(2)
        REQUIRE(value_at(4, {1, 2}, PairMode::Open, true) == 12.0 + 101.0 + 1002.0);
        REQUIRE(value_at(4, {1, 2}, PairMode::Cyclic, true) == 12.0 + 21.0 + 101.0 + 1002.0);
    }

    SECTION("general n") {
        const std::vector<int32_t> e = {2, 0, 3, 1};
        // d(2,0) + d(0,3) + d(3,1) = 20 + 3 + 31 = 54
        REQUIRE(value_at(4, e, PairMode::Open, false) == 54.0);
        REQUIRE(value_at(4, e, PairMode::Cyclic, false) == 54.0 + 12.0);  // + d(1,2)
        // + head(2) + tail(1) = 102 + 1001
        REQUIRE(value_at(4, e, PairMode::Open, true) == 54.0 + 102.0 + 1001.0);
        REQUIRE(value_at(4, e, PairMode::Cyclic, true) == 54.0 + 12.0 + 102.0 + 1001.0);
    }

    SECTION("Open is the default, so the pre-existing two-argument call is unchanged") {
        Model m;
        auto lv = m.list_var(4);
        int32_t node = m.pair_lambda_sum(lv, d);
        m.minimize(node);
        m.close();
        m.var_mut(vid(lv)).elements = {2, 0, 3, 1};
        full_evaluate(m);
        REQUIRE(m.node_value(node) == 54.0);
    }

    SECTION("an empty head or tail callable means no term") {
        Model m;
        auto lv = m.list_var(4);
        int32_t node = m.pair_lambda_sum(lv, d, head_cost, nullptr);
        m.minimize(node);
        m.close();
        m.var_mut(vid(lv)).elements = {2, 0, 3, 1};
        full_evaluate(m);
        REQUIRE(m.node_value(node) == 54.0 + 102.0);
    }
}

TEST_CASE("every list move keeps delta evaluation equal to a fresh full evaluation",
          "[dag][pair_lambda][moves]") {
    // One node per variant in ONE model, so a single move sequence exercises
    // all four and the fresh-copy comparison covers them together.
    Model m;
    auto lv = m.list_var(7, "seq");
    const int32_t open_node = m.pair_lambda_sum(lv, d);
    const int32_t cyclic_node = m.pair_lambda_sum(lv, d, PairMode::Cyclic);
    const int32_t open_ends = m.pair_lambda_sum(lv, d, head_cost, tail_cost);
    const int32_t cyclic_ends = m.pair_lambda_sum(lv, d, head_cost, tail_cost, PairMode::Cyclic);
    m.minimize(m.sum({open_node, cyclic_node, open_ends, cyclic_ends}));
    m.close();
    full_evaluate(m);

    RNG rng(1234);
    std::vector<std::string> seen;
    for (int round = 0; round < 60; ++round) {
        auto moves = generate_standard_moves(m.var(vid(lv)), rng);
        REQUIRE_FALSE(moves.empty());
        for (const Move& mv : moves) {
            seen.push_back(mv.move_type);
            std::vector<int32_t> changed = apply_move(m, mv);
            delta_evaluate(m, changed.data(), changed.size());

            // A fresh copy of the model at the same assignment, evaluated from
            // scratch: the reference the incremental value must match.
            Model fresh;
            auto flv = fresh.list_var(7, "seq");
            const int32_t f_open = fresh.pair_lambda_sum(flv, d);
            const int32_t f_cyclic = fresh.pair_lambda_sum(flv, d, PairMode::Cyclic);
            const int32_t f_open_ends = fresh.pair_lambda_sum(flv, d, head_cost, tail_cost);
            const int32_t f_cyclic_ends =
                fresh.pair_lambda_sum(flv, d, head_cost, tail_cost, PairMode::Cyclic);
            fresh.minimize(fresh.sum({f_open, f_cyclic, f_open_ends, f_cyclic_ends}));
            fresh.close();
            fresh.var_mut(vid(flv)).elements = m.var(vid(lv)).elements;
            full_evaluate(fresh);

            INFO("move " << mv.move_type);
            REQUIRE(m.node_value(open_node) == fresh.node_value(f_open));
            REQUIRE(m.node_value(cyclic_node) == fresh.node_value(f_cyclic));
            REQUIRE(m.node_value(open_ends) == fresh.node_value(f_open_ends));
            REQUIRE(m.node_value(cyclic_ends) == fresh.node_value(f_cyclic_ends));
        }
    }

    // The loop is only worth its cost if it really saw every generator.
    for (const char* want :
         {"list_swap", "list_2opt", "list_relocate", "list_or_opt_2", "list_or_opt_3"}) {
        INFO("generator " << want);
        REQUIRE(std::find(seen.begin(), seen.end(), want) != seen.end());
    }
}

TEST_CASE("a .cbls round-trip preserves the closing rule and the endpoint terms",
          "[io][pair_lambda]") {
    const std::vector<int32_t> elements = {2, 0, 3, 1};

    // head-only and tail-only are their own rows: the writer emits the two keys
    // through independent calls and the reader reads them through independent
    // calls, so a dropped or swapped one survives the both-endpoints row.
    struct Variant {
        const char* name;
        PairMode mode;
        bool head;
        bool tail;
        double expected;
    };
    const std::array<Variant, 6> variants = {{
        {"open", PairMode::Open, false, false, 54.0},
        {"cyclic", PairMode::Cyclic, false, false, 66.0},
        {"open+head", PairMode::Open, true, false, 54.0 + 102.0},
        {"open+tail", PairMode::Open, false, true, 54.0 + 1001.0},
        {"open+endpoints", PairMode::Open, true, true, 54.0 + 102.0 + 1001.0},
        {"cyclic+endpoints", PairMode::Cyclic, true, true, 66.0 + 102.0 + 1001.0},
    }};

    for (const Variant& v : variants) {
        INFO("variant " << v.name);
        Model m;
        auto lv = m.list_var(4, "seq");
        int32_t node = (v.head || v.tail) ? m.pair_lambda_sum(lv, d, v.head ? head_cost : nullptr,
                                                              v.tail ? tail_cost : nullptr, v.mode)
                                          : m.pair_lambda_sum(lv, d, v.mode);
        m.minimize(node);
        m.close();

        std::ostringstream out;
        save_model(m, out);
        const std::string saved = out.str();

        // An open chain with no endpoints is everything the format could say
        // before these variants existed, so it must still say exactly that.
        REQUIRE((saved.find("\"mode\"") != std::string::npos) == (v.mode == PairMode::Cyclic));
        REQUIRE((saved.find("\"head\"") != std::string::npos) == v.head);
        REQUIRE((saved.find("\"tail\"") != std::string::npos) == v.tail);

        std::istringstream in(saved);
        Model reloaded = load_model(in);
        reloaded.var_mut(0).elements = elements;
        full_evaluate(reloaded);
        REQUIRE(reloaded.node_value(reloaded.objective_id()) == v.expected);

        // Idempotent: a second save of the reloaded model is byte-identical.
        std::ostringstream out2;
        save_model(reloaded, out2);
        REQUIRE(out2.str() == saved);
    }
}

TEST_CASE("a .cbls round-trip of a Set child covers the whole universe", "[io][pair_lambda]") {
    // A Set's max_size is its cardinality bound, its universe_size the element
    // range. Tabulating by max_size wrote a matrix too narrow to index, and the
    // reload then threw on any element past it.
    Model m;
    auto sv = m.set_var(6, 2, 3, "chosen");
    int32_t node = m.pair_lambda_sum(sv, d, head_cost, tail_cost, PairMode::Cyclic);
    m.minimize(node);
    m.close();

    std::ostringstream out;
    save_model(m, out);
    std::istringstream in(out.str());
    Model reloaded = load_model(in);
    // Elements 4 and 5 are inside the universe but outside max_size == 3.
    reloaded.var_mut(0).elements = {5, 4};
    full_evaluate(reloaded);
    // d(5,4) + d(4,5) + head(5) + tail(4)
    REQUIRE(reloaded.node_value(reloaded.objective_id()) == 54.0 + 45.0 + 105.0 + 1004.0);
}

TEST_CASE("an unknown PairLambda mode is refused rather than read as open", "[io][pair_lambda]") {
    Model m;
    auto lv = m.list_var(3, "seq");
    m.minimize(m.pair_lambda_sum(lv, d, PairMode::Cyclic));
    m.close();
    std::ostringstream out;
    save_model(m, out);
    std::string corrupted = out.str();
    const auto at = corrupted.find("\"cyclic\"");
    REQUIRE(at != std::string::npos);
    corrupted.replace(at, std::string("\"cyclic\"").size(), "\"spiral\"");
    std::istringstream in(corrupted);
    REQUIRE_THROWS_AS(load_model(in), std::invalid_argument);
}

TEST_CASE("a frozen model shares the closing rule with its replicas", "[pair_lambda][share]") {
    // The side table is structure, not search state, so it belongs in
    // ModelStructure -- which a frozen model SHARES with every portfolio
    // replica rather than deep-copying (#157). A spec that had landed on the
    // per-worker side would leave each replica reading a default-constructed
    // entry, i.e. an open chain, while the master looked right.
    Model master;
    auto lv = master.list_var(4, "seq");
    const int32_t node = master.pair_lambda_sum(lv, d, head_cost, tail_cost, PairMode::Cyclic);
    master.minimize(node);
    master.close();
    master.var_mut(vid(lv)).elements = {2, 0, 3, 1};
    full_evaluate(master);
    master.freeze();
    REQUIRE(master.is_frozen());

    Model replica(master);
    REQUIRE(replica.is_frozen());
    replica.var_mut(vid(lv)).elements = {2, 0, 3, 1};
    full_evaluate(replica);
    REQUIRE(replica.node_value(node) == master.node_value(node));
    REQUIRE(replica.node_value(node) == 54.0 + 12.0 + 102.0 + 1001.0);

    // A replica's own assignment does not disturb the master's.
    replica.var_mut(vid(lv)).elements = {0, 1, 2, 3};
    full_evaluate(replica);
    REQUIRE(master.node_value(node) == 54.0 + 12.0 + 102.0 + 1001.0);
}

TEST_CASE("a cyclic pair_lambda_sum finds the optimal TSP tour", "[search][pair_lambda]") {
    // Seven cities on a circle, so the optimum is the convex-hull order and any
    // crossing costs more. The optimum is found by brute force in the test
    // rather than quoted, which keeps the assertion self-contained: 6! = 720
    // tours with city 0 pinned.
    constexpr int kCities = 7;
    std::vector<double> xs;
    std::vector<double> ys;
    for (int i = 0; i < kCities; ++i) {
        const double angle = 2.0 * 3.14159265358979323846 * i / kCities;
        xs.push_back(std::cos(angle));
        ys.push_back(std::sin(angle));
    }
    // Shuffle the city ids so the optimal tour is not the identity permutation,
    // which the List's initial order would otherwise hand the search for free.
    const std::vector<int> label = {0, 3, 6, 2, 5, 1, 4};
    auto dist = [&](int a, int b) {
        const double dx = xs[label[a]] - xs[label[b]];
        const double dy = ys[label[a]] - ys[label[b]];
        return std::sqrt((dx * dx) + (dy * dy));
    };

    std::vector<int> perm(kCities);
    std::iota(perm.begin(), perm.end(), 0);
    double best = std::numeric_limits<double>::infinity();
    // City 0 is pinned at the front: a cyclic tour is invariant under rotation,
    // so fixing it enumerates every distinct tour exactly once per direction.
    std::sort(perm.begin() + 1, perm.end());
    do {
        double total = 0.0;
        for (int i = 0; i < kCities; ++i) {
            total += dist(perm[i], perm[(i + 1) % kCities]);
        }
        best = std::min(best, total);
    } while (std::next_permutation(perm.begin() + 1, perm.end()));

    Model m;
    auto lv = m.list_var(kCities, "tour");
    int32_t length =
        m.pair_lambda_sum(lv, [&](int a, int b) { return dist(a, b); }, PairMode::Cyclic);
    m.minimize(length);
    m.close();

    // 100 000 iterations reaches the optimum on 10 of 10 seeds; three are
    // asserted here so the test is a property of the search rather than of one
    // lucky stream. The whole case costs well under a second -- a pure
    // objective model has no real row to satisfy, so a batch is cheap.
    for (uint64_t seed : {163U, 7U, 20240163U}) {
        INFO("seed " << seed);
        SearchResult r = solve_deterministic(m, 100000, seed);
        REQUIRE(r.feasible);
        REQUIRE_THAT(r.objective, Catch::Matchers::WithinAbs(best, 1e-9));
    }
}

File: tests/python/test_pair_lambda.py
"""pair_lambda_sum and the table-backed lambda forms, from Python (#163).

Three things are pinned here.

**Values.** Both functor forms and both table forms, against a hand
computation, including the short lists (n = 0, 1, 2) whose behaviour the C++
header defines rather than leaves to chance.

**No Python on the hot path.** `pair_table_sum`/`lambda_table_sum` exist
because a Python functor re-acquires the GIL once per pair per node evaluation,
which serialises every portfolio worker on the interpreter. `sys.setprofile`
counts the `call` events a solve produces: the functor form makes thousands,
the table form makes none at all. Without this test the table form could
silently degrade into a wrapper around a Python callable and still pass every
value assertion.

**Crashes, not exceptions.** An ndarray of the wrong shape and an element
outside the tabulated universe are both heap reads past the end of a C++
vector, which take the interpreter down rather than failing a test. Those run
in a CHILD interpreter, following `test_model_handles.py`.
"""

from __future__ import annotations

import os
import subprocess
import sys
from typing import TYPE_CHECKING

import _cbls_core as cbls
import numpy as np
import pytest

if TYPE_CHECKING:
    from collections.abc import Callable

CHILD_TIMEOUT_SECONDS = 20.0


def vid(handle: int) -> int:
    """Variable handle (negative, as the builders return) -> variable id."""
    return -(handle + 1)


def dist(a: int, b: int) -> float:
    """Asymmetric on purpose: a symmetric cost cannot tell a cyclic sum from twice an open one."""
    return 10.0 * a + b


DIST = np.array([[dist(a, b) for b in range(4)] for a in range(4)], dtype=np.float64)
HEAD = np.array([100.0 + e for e in range(4)], dtype=np.float64)
TAIL = np.array([1000.0 + e for e in range(4)], dtype=np.float64)


def _value(build: str, elements: list[int], cyclic: bool, endpoints: bool) -> float:
    """Evaluate one pair node at `elements`, which may be shorter than the list."""
    m = cbls.Model()
    lv = m.list_var(4, "seq")
    if build == "functor":
        node = m.pair_lambda_sum(
            lv,
            dist,
            cyclic=cyclic,
            head=(lambda e: 100.0 + e) if endpoints else None,
            tail=(lambda e: 1000.0 + e) if endpoints else None,
        )
    else:
        node = m.pair_table_sum(
            lv,
            DIST,
            cyclic=cyclic,
            head=HEAD if endpoints else None,
            tail=TAIL if endpoints else None,
        )
    m.minimize(node)
    m.close()
    m.var_mut(vid(lv)).elements = elements
    cbls.full_evaluate(m)
    return float(m.node_value(node))


@pytest.mark.parametrize("build", ["functor", "table"])
@pytest.mark.parametrize(
    ("elements", "cyclic", "endpoints", "expected"),
    [
        # n = 0: zero for every variant, endpoints included.
        ([], False, False, 0.0),
        ([], True, False, 0.0),
        ([], True, True, 0.0),
        # n = 1: no pair, and Cyclic adds none; endpoints both charge e_0.
        ([3], False, False, 0.0),
        ([3], True, False, 0.0),
        ([3], False, True, 103.0 + 1003.0),
        # n = 2: the one edge is traversed twice when cyclic.
        ([1, 2], False, False, 12.0),
        ([1, 2], True, False, 12.0 + 21.0),
        ([1, 2], False, True, 12.0 + 101.0 + 1002.0),
        # General n: d(2,0) + d(0,3) + d(3,1) = 20 + 3 + 31.
        ([2, 0, 3, 1], False, False, 54.0),
        ([2, 0, 3, 1], True, False, 54.0 + 12.0),
        ([2, 0, 3, 1], False, True, 54.0 + 102.0 + 1001.0),
        ([2, 0, 3, 1], True, True, 54.0 + 12.0 + 102.0 + 1001.0),
    ],
)
def test_pair_sum_matches_a_hand_computation(
    build: str, elements: list[int], cyclic: bool, endpoints: bool, expected: float
) -> None:
    got = _value(build, elements, cyclic=cyclic, endpoints=endpoints)
    assert got == expected


def test_the_functor_and_table_forms_agree_on_every_variant() -> None:
    elements = [3, 1, 0, 2]
    for cyclic in (False, True):
        for endpoints in (False, True):
            functor = _value("functor", elements, cyclic=cyclic, endpoints=endpoints)
            table = _value("table", elements, cyclic=cyclic, endpoints=endpoints)
            assert functor == table, (cyclic, endpoints)


def test_lambda_table_sum_matches_lambda_sum() -> None:
    table = np.array([1.0, 2.5, 4.0, 8.0], dtype=np.float64)
    m = cbls.Model()
    lv = m.list_var(4)
    by_table = m.lambda_table_sum(lv, table)
    by_functor = m.lambda_sum(lv, lambda e: float(table[e]))
    m.minimize(m.sum([by_table, by_functor]))
    m.close()
    m.var_mut(vid(lv)).elements = [2, 0, 3]
    cbls.full_evaluate(m)
    assert m.node_value(by_table) == 4.0 + 1.0 + 8.0
    assert m.node_value(by_table) == m.node_value(by_functor)


def test_a_table_is_copied_so_mutating_the_caller_s_array_does_not_change_the_model() -> None:
    table = np.array([1.0, 2.0, 3.0, 4.0], dtype=np.float64)
    m = cbls.Model()
    lv = m.list_var(4)
    node = m.lambda_table_sum(lv, table)
    m.minimize(node)
    m.close()
    table[:] = 99.0
    m.var_mut(vid(lv)).elements = [0, 1, 2, 3]
    cbls.full_evaluate(m)
    assert m.node_value(node) == 10.0


def _count_python_calls_during_solve(use_table: bool) -> tuple[int, float]:
    """Solve a small tour model and count the Python-level calls the search made."""
    cities = 9
    matrix = np.array(
        [[dist(a, b) for b in range(cities)] for a in range(cities)], dtype=np.float64
    )
    m = cbls.Model()
    lv = m.list_var(cities, "tour")
    if use_table:
        node = m.pair_table_sum(lv, matrix, cyclic=True)
    else:
        node = m.pair_lambda_sum(lv, dist, cyclic=True)
    m.minimize(node)
    m.close()

    config = cbls.SearchConfig()
    config.max_iterations = 50000

    calls = 0

    def profiler(frame: object, event: str, arg: object) -> None:
        nonlocal calls
        if event == "call":
            calls += 1

    sys.setprofile(profiler)
    try:
        # time_limit = 0 -> the iteration budget alone bounds the run, so the
        # two arms do the same amount of work and the counts are comparable.
        result = cbls.solve(m, 0.0, 42, config=config)
    finally:
        sys.setprofile(None)
    assert node >= 0
    return calls, float(result.objective)


def test_the_table_form_makes_no_python_calls_during_solve() -> None:
    functor_calls, functor_objective = _count_python_calls_during_solve(use_table=False)
    table_calls, table_objective = _count_python_calls_during_solve(use_table=True)

    # The functor arm is the control: if it did not call back into Python
    # either, the counter is measuring nothing.
    assert functor_calls > 1000, functor_calls
    assert table_calls == 0, table_calls
    # Same cost function, so the same trajectory and the same answer.
    assert table_objective == functor_objective


def _run_scenario(name: str) -> subprocess.CompletedProcess[str]:
    """Run one `__main__` scenario in a child interpreter, under a hard deadline."""
    env = dict(os.environ)
    # No conftest.py in the child to put the build directory on the path.
    env["PYTHONPATH"] = os.pathsep.join(p for p in sys.path if p)
    env["PYTHONOPTIMIZE"] = "0"
    return subprocess.run(
        [sys.executable, os.path.abspath(__file__), name],
        capture_output=True,
        text=True,
        timeout=CHILD_TIMEOUT_SECONDS,
        check=False,
        env=env,
    )


@pytest.mark.parametrize("scenario", ["bad_table", "out_of_universe_element"])
def test_a_bad_table_or_element_raises_instead_of_reading_past_the_table(scenario: str) -> None:
    proc = _run_scenario(scenario)
    assert proc.returncode == 0, (
        f"child exited {proc.returncode} (negative = killed by that signal)\n"
        f"stdout:\n{proc.stdout}\nstderr:\n{proc.stderr}"
    )
    assert proc.stdout.strip().endswith("OK"), proc.stdout


def _scenario_bad_table() -> None:
    """Every way a mis-sized or mis-typed table reaches the engine is refused."""
    m = cbls.Model()
    lv = m.list_var(4)
    sv = m.set_var(5)
    b = m.bool_var()

    cases: list[tuple[str, Callable[[], object], type[Exception]]] = [
        ("short matrix", lambda: m.pair_table_sum(lv, np.zeros((3, 3))), ValueError),
        ("long matrix", lambda: m.pair_table_sum(lv, np.zeros((5, 5))), ValueError),
        ("non-square", lambda: m.pair_table_sum(lv, np.zeros((4, 3))), ValueError),
        ("1-D dist", lambda: m.pair_table_sum(lv, np.zeros(4)), TypeError),
        ("short head", lambda: m.pair_table_sum(lv, DIST, head=np.zeros(3)), ValueError),
        ("short tail", lambda: m.pair_table_sum(lv, DIST, tail=np.zeros(9)), ValueError),
        ("set universe", lambda: m.pair_table_sum(sv, DIST), ValueError),
        ("scalar var", lambda: m.pair_table_sum(b, DIST), ValueError),
        ("node handle", lambda: m.pair_table_sum(0, DIST), ValueError),
        ("stale handle", lambda: m.pair_table_sum(-100, DIST), IndexError),
        ("short lambda table", lambda: m.lambda_table_sum(lv, np.zeros(3)), ValueError),
        ("2-D lambda table", lambda: m.lambda_table_sum(lv, np.zeros((4, 4))), TypeError),
    ]
    for name, call, expected in cases:
        nodes_before = m.num_nodes()
        try:
            call()
        except expected:
            pass
        else:
            raise AssertionError(f"{name} was accepted")
        # A refused node leaves no trace, so the model is still usable.
        assert m.num_nodes() == nodes_before, name

    # The right shapes still work, on both variable types.
    m.pair_table_sum(lv, DIST, cyclic=True, head=HEAD, tail=TAIL)
    m.lambda_table_sum(sv, np.zeros(5))
    print("OK")


def _scenario_out_of_universe_element() -> None:
    """`Variable.elements` is writable from Python, so the lookup stays range-checked."""
    m = cbls.Model()
    lv = m.list_var(4, "seq")
    node = m.pair_table_sum(lv, DIST, cyclic=True, head=HEAD, tail=TAIL)
    m.minimize(node)
    m.close()

    for bad in ([0, 1, 9], [0, -3], [77]):
        m.var_mut(vid(lv)).elements = bad
        try:
            cbls.full_evaluate(m)
        except IndexError:
            continue
        raise AssertionError(f"elements {bad} were read past the table")

    # And the in-universe assignment still evaluates.
    m.var_mut(vid(lv)).elements = [0, 1, 2, 3]
    cbls.full_evaluate(m)
    assert m.node_value(node) == 1.0 + 12.0 + 23.0 + 30.0 + 100.0 + 1003.0
    print("OK")


if __name__ == "__main__":
    scenario = sys.argv[1]
    if scenario == "bad_table":
        _scenario_bad_table()
    elif scenario == "out_of_universe_element":
        _scenario_out_of_universe_element()
    else:
        raise SystemExit(f"unknown scenario {scenario}")
